explorer/interpreter/value.h

```diff
@@ -379,44 +381,48 @@ class NominalClassValue : public Value {
   Nonnull<const NominalClassValue** const> class_value_ptr_;
 };
 
+// Common implementation of alternative values and alternative constructors.
```

Comment:
This comment strongly suggests to me that inheritance is the wrong model of code reuse here, because it suggests that this class is just a container for some shared implementation details that have been factored out, not an entity that's genuinely meaningful in its own right.

This isn't just a theoretical point. On a first pass through this code, I thought the idea was that one of the derived class represents values whose alternative was declared with parentheses, and the other represents values whose alternative was declared without parentheses, and I spent several minutes leaving comments on that misguided basis. I think this base class is what caused the confusion, because it encouraged me to look for an interpretation where `AlternativeValueBase` represents a semantically meaningful category.

Composition is generally preferable to inheritance when you want to reuse code and don't want to model an is-a relationship, but in this case there's so little code that it may be best to just duplicate it.

Reply:
#2606. If only we had mixins :)